hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java

```diff
@@ -207,40 +210,42 @@ private void markClosed() {
     /**
      * Fetches the next page.
      */
-    private SqlPage fetch() {
+    private SqlPage fetch(long timeout, TimeUnit timeUnit) {
         synchronized (mux) {
-            // Re-throw previously logged error on successive fetch attempts.
             if (fetch != null) {
-                assert fetch.getError() != null;
-
-                throw wrap(fetch.getError());
+                if (fetch.getError() != null) {
+                    // Re-throw previously logged error on successive fetch attempts.
+                    throw wrap(fetch.getError());
+                }
+            } else {
+                // Initiate the fetch.
+                fetch = new SqlFetchResult();
+                service.fetchAsync(connection, queryId, cursorBufferSize, this);
             }
 
-            // Initiate the fetch.
-            fetch = new SqlFetchResult();
-
-            service.fetchAsync(connection, queryId, cursorBufferSize, this);
-
             // Await the response.
-            while (fetch.isPending()) {
+            long waitNanos = timeUnit.toNanos(timeout);
+            while (fetch.isPending() && waitNanos > 0) {
                 try {
-                    mux.wait();
+                    long startNanos = System.nanoTime();
+                    TimeUnit.NANOSECONDS.timedWait(mux, timeUnit.toNanos(waitNanos));
```

Comment:
This is duplicate unit conversion, it was already converted on line 227. But it works because the `timeUnit` argument to this method is always NANOSECONDS. We should fix both issues.